generator/src/test/java/com/scottlogic/deg/generator/builders/ConstraintChainBuilder.java

```diff
@@ -0,0 +1,83 @@
+package com.scottlogic.deg.generator.builders;
+
+import com.scottlogic.deg.generator.Field;
+import com.scottlogic.deg.generator.constraints.Constraint;
+import com.scottlogic.deg.generator.constraints.atomic.AtomicConstraint;
+import com.scottlogic.deg.generator.constraints.atomic.IsGreaterThanConstantConstraint;
+import com.scottlogic.deg.generator.constraints.atomic.IsLessThanConstantConstraint;
+import com.scottlogic.deg.generator.constraints.atomic.ViolatedAtomicConstraint;
+import com.scottlogic.deg.generator.constraints.grammatical.AndConstraint;
+import com.scottlogic.deg.generator.constraints.grammatical.OrConstraint;
+
+import java.util.ArrayList;
+import java.util.Collection;
+
+/**
+ * Defines a builder for a class that can contain constraints.
+ */
+public abstract class ConstraintChainBuilder<T> {
+    protected Collection<Constraint> constraints = new ArrayList<>();
+    private Constraint constraintToAdd;
+
+    public abstract T build();
```

Comment:
Should this be an implementation such as:
```suggestion
    public abstract T build(){
       saveConstraint();  //to save the need to call saveConstraint() in each implementation
       return buildConstraint(); //the abstract method, albeit renamed (can be marked package private too?)
    }
    
    abstract T buildConstraint();
```

Therefore `AndBuilder` would look like:
```
    AndConstraint buildConstraint() {
        return new AndConstraint(constraints);
    }
```

Reply:
yeah I like this